### PR TITLE
display trace data in table 

### PR DESCRIPTION
**Pull request by aeolusheath**
add new way to display trace data:

image

### CODE REVIEW

**Comment by wu-sheng** [comment]
Table should be optional view, put it at the last one

File: src/views/components/trace/trace-table.vue [hunk 1]
```diff
@@ -78,6 +78,12 @@ export default class Home extends Vue {
       this.loading = false;
     });
   }
+  // @Watch('rocketTrace.traceList')
+  // onListChanged(val: any, oldVal: any) {
+  //   if (this.rocketTrace && this.rocketTrace.traceList.length > 0) {
+  //     this.selectTrace(this.rocketTrace.traceList[0])
+  //   }
+  // }
 }
```
Comment (peng-yongsheng) [inline src/views/components/trace/trace-table.vue]: Delete them.

File: src/views/components/trace/trace-chart-table/trace-item.vue [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    execTime() {
+      const {data} = this;
+      return  (data.endTime - data.startTime) ? (data.endTime - data.startTime) : 0;
+    },
+    outterPercent() {
+      if (this.data.level == 1) {
+        return '100%';
+      } else {
+        let data = this.data
+        // let exec = ((data.endTime - data.startTime) && data.children && data.children.length > 0 ) ? (data.endTime - data.startTime) : 0
```
Comment (peng-yongsheng) [inline src/views/components/trace/trace-chart-table/trace-item.vue]: Delete it.

File: src/views/components/trace/trace-chart-table/trace-item.vue [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        </div>
+        <div class="start-time">
+          {{formatTime(data.startTime)}}
+        </div>
+        <div class="gap">
+          0
+        </div>
+        <div class="exec-ms">
+         {{(data.endTime - data.startTime)?(data.endTime - data.startTime) : '0'}} 
+         <!-- {{( (data.endTime - data.startTime) && data.children && data.children.length > 0 ) || (data.level == 1) ?(data.endTime - data.startTime) + '' : '0'}}  -->
```
Comment (peng-yongsheng) [inline src/views/components/trace/trace-chart-table/trace-item.vue]: Delete it.

**Comment by wu-sheng** [comment]
As this is too closed to pinpoint UI, I am asking the pinpoint team to verify about IP check. Let's at least wait 72 hours 

https://twitter.com/wusheng1108/status/1156844539929227264

File: src/main.ts [hunk 4]
```diff
@@ -66,5 +66,8 @@ new Vue({
   i18n,
   router,
   store,
+  data: {
+    eventHub: new Vue(),
```
Comment (TinyAllen) [inline src/main.ts]: There is already vuex that should not use bus for relay

File: src/utils/tooltip.ts [hunk 5]
```diff
@@ -110,6 +110,11 @@ export default {
 
     const $inner = document.createElement('div');
     $inner.setAttribute('class', 'rk-tooltip-inner');
+
+    if (binding.modifiers.inTraceTable) {
+      $inner.setAttribute('class', 'rk-tooltip-inner trace-table-tooltip-inner');
```
Comment (TinyAllen) [inline src/utils/tooltip.ts]: Utils files need to be shared and cannot be specific to specific files，please use namespace

File: src/views/components/trace/trace-chart-table/trace-container.vue [hunk 6]
```diff
@@ -0,0 +1,110 @@
+<template>
```
Comment (TinyAllen) [inline src/views/components/trace/trace-chart-table/trace-container.vue]: image

warning need fix

File: src/views/components/trace/trace-chart-table/trace-item.vue [hunk 7]
```diff
@@ -0,0 +1,210 @@
+<template>
```
Comment (TinyAllen) [inline src/views/components/trace/trace-chart-table/trace-item.vue]: image

warning need fix

File: src/views/components/trace/trace-chart-table/trace-item.vue [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+    <div v-show="data.children && data.children.length > 0 && displayChildren" class="children-trace">
+        <item v-for="(item, index) in data.children" :key="index" :data="item"> </item>
+    </div>
+  </div>
+
+</template>
+<style lang="scss" scoped>
+  @for $i from 0 through 20 {
+    .method.level#{$i} {
+      text-indent: #{$i * 10}px;
```
Comment (TinyAllen) [inline src/views/components/trace/trace-chart-table/trace-item.vue]: Only 20-tier structure is supported here. Use DOM recursive nesting to achieve infinite tier

File: src/views/components/trace/trace-detail-chart-table.vue [hunk 9]
```diff
@@ -0,0 +1,148 @@
+<template>
```
Comment (TinyAllen) [inline src/views/components/trace/trace-detail-chart-table.vue]: image

warning need fixed

File: src/views/components/trace/trace-detail-chart-table.vue [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+      list: [],
+      currentSpan: []      
+    }
+  },
+  methods: {
+    highlight () {
+      this.selected = true;
+    },
+    // 给增加层级关系
+    formatData (arr, level = 1, totalExec = null) {
```
Comment (TinyAllen) [inline src/views/components/trace/trace-detail-chart-table.vue]: image
image

# very serious
The chart is incorrectly displayed.

File: src/views/components/trace/trace-detail.vue [hunk 11]
```diff
@@ -29,22 +29,32 @@
...
+
+            
       <div class="rk-tag mr-5">{{this.$t('start')}}</div><span class="mr-15 sm">{{parseInt(current.start) | dateformat}}</span>
       <div class="rk-tag mr-5">{{this.$t('duration')}}</div><span class="mr-15 sm">{{current.duration}} ms</span>
       <div class="rk-tag mr-5">{{this.$t('spans')}}</div><span class="sm">{{spans.length}}</span>
     </div>
-    <TraceDetailChartList v-if="mode&&current.endpointNames" :data="spans" :traceId="current.traceIds[0]"/>
-    <TraceDetailChartTree v-if="!mode&&current.endpointNames" :data="spans" :traceId="current.traceIds[0]"/>
+    <TraceDetailChartList v-if="displayMode == 'list'&&current.endpointNames" :data="spans" :traceId="current.traceIds[0]"/>
+    <TraceDetailChartTree v-if="displayMode == 'tree'&&current.endpointNames" :data="spans" :traceId="current.traceIds[0]"/>
```
Comment (TinyAllen) [inline src/views/components/trace/trace-detail.vue]: image

Incorrect selection status

File: src/views/components/trace/trace-table.vue [hunk 12]
```diff
@@ -49,7 +49,7 @@
 </template>
 
 <script lang="ts">
-import { Component, Vue, Prop } from 'vue-property-decorator';
+import { Component, Vue, Prop, Watch } from 'vue-property-decorator';
```
Comment (TinyAllen) [inline src/views/components/trace/trace-table.vue]: Avoid using Watch ，Use computer instead。

File: src/views/components/trace/trace-detail.vue [hunk 13]
```diff
@@ -29,22 +29,32 @@
           <option v-for="i in current.traceIds" :value="i" :key="i">{{i}}</option>
         </select>
       </div>
-      <a class="rk-btn sm r" :class="{'ghost':mode}" @click="mode = false">
+
```
Comment (TinyAllen) [inline src/views/components/trace/trace-detail.vue]: image
The color is incurrect, active is blue and unactive is dark

File: src/views/components/trace/trace-chart-table/trace-container.vue [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+        </div>
+      </div>
+
+      <slot> </slot>
+    </div>
+</template>
+<style lang="scss" scoped>
+.trace {
+    font-size: 12px;
+    margin: 5px;
```
Comment (TinyAllen) [inline src/views/components/trace/trace-chart-table/trace-container.vue]: Remove this line margin:5px

File: src/views/components/trace/trace-chart-table/trace-container.vue [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+    height: 100%;
+    overflow: auto;
+  }
+  .trace-header {
+    display: flex;
+    border: 2px solid #ddd;
```
Comment (TinyAllen) [inline src/views/components/trace/trace-chart-table/trace-container.vue]: Remove this line border: 2px solid #ddd;

File: src/views/components/trace/trace-chart-table/trace-container.vue [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+    border-left: 0;
+    border-right: 0;
+  }
+
+  .trace-header div {
+    padding: 0 4px;
+    border: 1px solid transparent;
+    border-right: 1px dotted silver;
+    line-height: 30px;
+    background-color: aliceblue;
```
Comment (TinyAllen) [inline src/views/components/trace/trace-chart-table/trace-container.vue]: change into background-color: #f3f4f9;

File: src/views/components/trace/trace-chart-table/trace-container.vue [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+    border: 1px solid transparent;
+    border-right: 1px dotted silver;
+    overflow: hidden;
+    line-height: 30px;
+    overflow:hidden; text-overflow:ellipsis; white-space:nowrap        
+  }
+
+
+
+  .trace-item {
```
Comment (TinyAllen) [inline src/views/components/trace/trace-chart-table/trace-container.vue]: remove this already exited in trace-item

File: src/views/components/trace/trace-chart-table/trace-container.vue [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+    display: flex;
+  }
+  .trace-item.selected {
```
Comment (TinyAllen) [inline src/views/components/trace/trace-chart-table/trace-container.vue]: remove trace-item in Trace-container.vue, this has already exited in trace-item

File: src/views/components/trace/trace-chart-table/trace-item.vue [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+      </div>      
+    <div v-show="data.children && data.children.length > 0 && displayChildren" class="children-trace">
+        <item v-for="(item, index) in data.children" :key="index" :data="item"> </item>
+    </div>
+  </div>
+
+</template>
+<style lang="scss" scoped>
+  .trace-item.level0{
+      background: #DFF0D8;
```
Comment (TinyAllen) [inline src/views/components/trace/trace-chart-table/trace-item.vue]: change into background: rgba(0, 0, 0, 0.04);

File: src/views/components/trace/trace-chart-table/trace-item.vue [hunk 20]
```diff
@@ -0,0 +1,207 @@
...
+      color: #1469EB;
```
Comment (TinyAllen) [inline src/views/components/trace/trace-chart-table/trace-item.vue]: change into color: #448dfe

File: src/views/components/trace/trace-chart-table/trace-item.vue [hunk 21]
```diff
@@ -0,0 +1,207 @@
...
+      &:hover {
+        background: #DFF0D8;
```
Comment (TinyAllen) [inline src/views/components/trace/trace-chart-table/trace-item.vue]: change into background: rgba(0, 0, 0, 0.04);

File: src/views/components/trace/trace-chart-table/trace-item.vue [hunk 22]
```diff
@@ -0,0 +1,207 @@
...
+        color: #1469EB;
```
Comment (TinyAllen) [inline src/views/components/trace/trace-chart-table/trace-item.vue]: change into color: #448dfe

File: src/views/components/trace/trace-chart-table/trace-item.vue [hunk 23]
```diff
@@ -0,0 +1,207 @@
...
+      position: absolute;
+      content: '';
+      width: 5px;
+      height: 100%;
+      background: #cb7686;
+      left: 0;
+    }
+  }
+  .trace-item.selected {
+    background: beige;
```
Comment (TinyAllen) [inline src/views/components/trace/trace-chart-table/trace-item.vue]: change into background: rgba(0, 0, 0, 0.04);

File: src/views/components/trace/trace-chart-table/trace-item.vue [hunk 24]
```diff
@@ -0,0 +1,207 @@
...
+  }
+
+  .trace-item:not(.level0):hover {
+    background: beige;
```
Comment (TinyAllen) [inline src/views/components/trace/trace-chart-table/trace-item.vue]: change into background: rgba(0, 0, 0, 0.04);

File: src/views/components/trace/trace-chart-table/trace-item.vue [hunk 25]
```diff
@@ -0,0 +1,207 @@
...
+    border-top: 1px solid white;
```
Comment (TinyAllen) [inline src/views/components/trace/trace-chart-table/trace-item.vue]: remove this line

File: src/views/components/trace/trace-chart-table/trace-container.vue (hunk 6 above)
Comment (TinyAllen) [reply src/views/components/trace/trace-chart-table/trace-container.vue]: use <script lang="js"> not <script type="text/javascript"> to fix it

File: src/views/components/trace/trace-chart-table/trace-item.vue (hunk 7 above)
Comment (TinyAllen) [reply src/views/components/trace/trace-chart-table/trace-item.vue]: use <script lang='js'>  not<script type="text/javascript"> to fix

File: src/views/components/trace/trace-detail-chart-table.vue (hunk 9 above)
Comment (TinyAllen) [reply src/views/components/trace/trace-detail-chart-table.vue]: use <script lang='js'> not<script type="text/javascript"> to fix

**Comment by peng-yongsheng** [comment]
> As this is too closed to pinpoint UI, I am asking the pinpoint team to verify about IP check. Let's at least wait 72 hours
> 
> https://twitter.com/wusheng1108/status/1156844539929227264

72 hours have passed, there is no response from the Pinpoint team. I think they're fine with this.

@TinyAllen Hello, is there any question with this PR, @aeolusheath told me he had fixed all the issues you provide.

**Comment by TinyAllen** [comment]
@peng-yongsheng  @aeolusheath  l have check all the conversation, some style have not change yet, I've given the revision plan. Not to copy the same style from pinpoint.

**Comment by aeolusheath** [comment]
> @peng-yongsheng @aeolusheath l have check all the conversation, some style have not change yet, I've given the revision plan. Not to copy the same style from pinpoint.

Hi, could u please check the code again?  I'm sure I have modified the code as your advices  on the commit `bbbc074`.

And I adjust the style again on the lastest commit `015daf1`.

thanks for review the codes  🍻 💙

**Comment by wu-sheng** [comment]
@aeolusheath Please submit a submodule update pull request to the main repo, in order to sync this to the SkyWalking master.

**Pull request merged**